File: leetcode_java/src/MyHashMap.java
//Q706
public class MyHashMap {

	ListNode[] bucket = new ListNode[10000];
	/** Initialize your data structure here. */
    
	private int getBucketIndex(int key){	
		return Integer.hashCode(key) % bucket.length;
	}
	
    /** value will always be non-negative. */
    public void put(int k, int v) {
    	int index = getBucketIndex(k);	
     	if (bucket[index] == null) {
     		bucket[index] = new ListNode(k, v);
     	} else if (bucket[index].key == k){
     		bucket[index].value = v;
     	} else {
             ListNode pre = bucket[index];
     		while(pre.next != null && pre.next.key != k) {
             	pre = pre.next;
             }
             if (pre.next != null) {
             	pre.next.value = v;
             }else {
             	pre.next = new ListNode(k, v);
             }
     	}
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    public int get(int k) {
    	int index = getBucketIndex(k);
    	ListNode cur = bucket[index];
        while(cur != null && cur.key != k) {
        	cur = cur.next;
        }
        if (cur != null) {
        	return cur.value;
        }else {
        	return -1;
        }
//        printMap();
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    public void remove(int k) {
    	int index = getBucketIndex(k);
    	ListNode pre = bucket[index];
    	if (pre != null && pre.key == k) {
    		bucket[index] = pre.next;
    	} else {
            while(pre != null && pre.next != null && pre.next.key != k) {
            	pre = pre.next;
            }
            if (pre != null && pre.next != null) {
            	pre.next = pre.next.next;
            }
    	}
    }

//    public void printMap() {
////    	ListNode cur = head;
//    	System.out.println("MyMap");
//    	while (cur!=null) {
//    		System.out.println("<"+cur.key+" , "+cur.value+">");
//    		cur = cur.next;
//    	}
//    }
    
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		MyHashMap map = new MyHashMap();
		map.put(1, 1);
		map.put(2, 2);
		map.put(2, 3);
	}

	class ListNode {
		int key;
		int value;
		ListNode next;
		
		ListNode(int myk, int myv){
			key = myk;
			value = myv;
			next = null;
		}
	}
}

File: leetcode_python/MyHashMap.py
# Q706
class MyHashMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.keys = []
        self.values = []


    def put(self, key: int, value: int) -> None:
        """
        value will always be non-negative.
        """
        if key in self.keys:
            i = self.keys.index(key)
            self.values[i] = value
        else:
            self.keys.append(key)
            self.values.append(value)



    def get(self, key: int) -> int:
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        """
        if key in self.keys:
            return self.values[self.keys.index(key)]
        else:
            return -1


    def remove(self, key: int) -> None:
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        """
        if key in self.keys:
            i = self.keys.index(key)
            del self.keys[i]
            del self.values[i]



# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)
